Add an error guard around app

samelhusseini opened the pull request:
React no has support for error boundaries: https://reactjs.org/docs/error-boundaries.html

This would mean that if an unhandled error occurs somewhere in the React components, we can display a message like so, and get users to reload the page, log the error etc. 

<img width="770" alt="screen shot 2018-04-26 at 2 02 21 pm" src="https://user-images.githubusercontent.com/16690124/39323558-eb94a0d8-495a-11e8-9875-5095d6bbb88c.png">

Alternatively, we can just reload the page for them but I feel like that's a little intrusive. 

thoughts?

pelikhan commented:
We need to reload the page. There is no reload button on iOS/Windows app.

_____________________________
From: Sam El-Husseini <notifications@github.com>
Sent: Thursday, April 26, 2018 2:06 PM
Subject: [Microsoft/pxt] Add an error guard around the entire application for unhandled errors. (#4215)
To: Microsoft/pxt <pxt@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>



React no has support for error boundaries: https://reactjs.org/docs/error-boundaries.html<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Freactjs.org%2Fdocs%2Ferror-boundaries.html&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877541947&sdata=LHGhhyEN3FqXNCZpElUaqfrD2z3aeeO7qhHoHilluXs%3D&reserved=0>

This would mean that if an unhandled error occurs somewhere in the application, we can display a message like so, and get users to reload the page, log the error etc.

[screen shot 2018-04-26 at 2 02 21 pm]<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F16690124%2F39323558-eb94a0d8-495a-11e8-9875-5095d6bbb88c.png&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877551948&sdata=To1nhVUF59tVLmPJcru69svSZRnK4LL0fD6d1rnxkJI%3D&reserved=0>

Alternatively, we can just reload the page for them but I feel like that's a little intrusive.

thoughts?

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/Microsoft/pxt/pull/4215<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fpxt%2Fpull%2F4215&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877551948&sdata=lQM95eL4snnnMSyn5XEqjbO6h9RWXM3Tr67%2BKbHGb2A%3D&reserved=0>

Commit Summary

  *   Add an error guard around the entire application for unhandled errors.

File Changes

  *   M pxteditor/editor.ts<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fpxt%2Fpull%2F4215%2Ffiles%23diff-0&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877561952&sdata=vB2m7tothU24tJgMB7SsYuSJpZO7lq9LKRah4x%2FucqQ%3D&reserved=0> (1)
  *   M webapp/src/app.tsx<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fpxt%2Fpull%2F4215%2Ffiles%23diff-1&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877561952&sdata=QYKHhnxJbM1GWTLXADYdOZtamFgBp81bu55Cp0qI1ks%3D&reserved=0> (20)

Patch Links:

  *   https://github.com/Microsoft/pxt/pull/4215.patch<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fpxt%2Fpull%2F4215.patch&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877571966&sdata=wGwGaz5TqFiUwbDg9CcbAsveUjWUJqvYDeX932y7944%3D&reserved=0>
  *   https://github.com/Microsoft/pxt/pull/4215.diff<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fpxt%2Fpull%2F4215.diff&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877571966&sdata=%2FTg6VTwNEWi4caXJNP57F%2FeTveiSWF5gxK9eBM0rnjo%3D&reserved=0>

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fpxt%2Fpull%2F4215&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877581980&sdata=6o1jaP3i91cvkTGChO3G3Td4hiOcQtF58zCPGmiH0IA%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAD-4KYhGy5A2c6jYDclR0C74CPdBwKIvks5tsgyggaJpZM4TmgYH&data=02%7C01%7Cjhalleux%40microsoft.com%7C890d1441709640ab734c08d5aba06d5f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603627877581980&sdata=tXNFze3nX5trTvv0%2BamGq9wfLtXSkB59MB%2F0hc2Xkh0%3D&reserved=0>.

samelhusseini commented:
Maybe still display the message, and reload after 1-2 seconds?

samelhusseini commented:
It would be helpedful for our testers / users to report an error by saying, I got that message.

samelhusseini commented:
Added a reload after 2 seconds. 

Also changed text to: 
"Oops :(:(:("
"This doesn't happen often, but looks like we've run into a problem.."
"Refreshing your page..."

pelikhan commented on webapp/src/app.tsx:
```diff
@@ -437,6 +437,19 @@ export class ProjectView
 
+    // Add an error guard for the entire application
+    componentDidCatch(error: any, info: any) {
+        core.killLoadingQueue();
```
don't do anything if hasError is already set.

samelhusseini replied:
I'm logging a criticalError event. I believe app insights already logs those anyway but it'd be good to know about the frequency of this event and what component is problematic. 

If react catches a number of exceptions we probably want to log all of them, what's that harm in doing this for all errors? (stop loading is harmless if done more than once, and same with set state) and it's going to all reload in 2 seconds anyway

pelikhan commented on webapp/src/app.tsx:
```diff
@@ -1760,6 +1773,17 @@ export class ProjectView
+            return <div id="root" className="ui middle aligned center aligned grid" style={{ height: '100%', alignItems: 'center' }}>
+                <div className="ui raised segment inverted purple">
+                    <h2>{lf("Oops") + "  😞😞😞"}</h2>
+                    {lf("This doesn't happen often, but looks like we've run into a problem..")}
```
Let's keep the wording simpler. Something like "Sorry we detected a problem and we will reload the editor in a few seconds." Or something better but let's not make claims about frequency.

samelhusseini replied:
Haha fair enough, I was trying to be cute

riknoll commented:
I would argue that a continuously refreshing-site looks much more broken then an editor that does nothing. If we do the auto-reload I would feel more comfortable if we added logic to limit it to one reload.

samelhusseini commented:
Yeah I'm on the fence here. I agree one reload is the ideal, as most issues we can recover from one reload.

samelhusseini commented:
removing auto-reload for now. I'm more interested in the telemetry, and we can choose to do something about it in the future.

pelikhan commented:
We used this strategy in TouchDevelop very efficiently. You could drop a time stamp in local storage to detect reload loops.

_____________________________
From: Sam El-Husseini <notifications@github.com>
Sent: Thursday, April 26, 2018 6:28 PM
Subject: [Microsoft/pxt] Add an error guard around the entire application for unhandled errors. (#4215)
To: Microsoft/pxt <pxt@noreply.github.com>
Cc: Peli de Halleux <jhalleux@microsoft.com>, Review requested <review_requested@noreply.github.com>



removing auto-reload for now. I'm more interested in the telemetry, and we can choose to do something about it in the future.

—
You are receiving this because your review was requested.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fpxt%2Fpull%2F4215%23issuecomment-384808332&data=02%7C01%7Cjhalleux%40microsoft.com%7C76ada38553af48012ed408d5abc517f5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603785355203075&sdata=tMFImPm3zsKihGwD7h9b7RGiW%2FuczMpyvpTx8CFhN14%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAD-4KXgQG5loWdiWfWcV7sngj9mBQhH5ks5tskojgaJpZM4TmgYH&data=02%7C01%7Cjhalleux%40microsoft.com%7C76ada38553af48012ed408d5abc517f5%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636603785355203075&sdata=GTnyoMeqgaRMpX7iN90urqRuQllxKoU%2BhM00UWvA20s%3D&reserved=0>.

riknoll commented:
+1 to dropping a timestamp

samelhusseini commented:
If you use a localstorage flag, say you set it and app reloads (avoiding a loop). When do you reset the flag? ie: say the app is now working right again? 
If you do it at a point where the app hasn't gotten to during it's startup, and then it crashes afterwards, you're still stuck in a reload loop. Let me know if you're not following

pelikhan commented:
If you don't do auto-refresh, you need to have a button that says "refresh" because there is no refresh button in some of our clients like the Windows app.

samelhusseini commented:
Yeah will do autorefresh and just set a timestamp.

pelikhan commented on webapp/src/simulator.ts:
```diff
@@ -33,6 +33,7 @@ export function setTranslations(translations: pxt.Map<string>) {
 }
 
 export function init(root: HTMLElement, cfg: SimulatorConfig) {
+    if (!root) return;
```
any reason why we need this?

samelhusseini replied:
Yes, boardview might not exist. 
It was throwing in cases where checkSession throws (due to an invalid session).

samelhusseini commented on webapp/src/app.tsx:
```diff
@@ -437,6 +437,25 @@ export class ProjectView
+        pxt.tickEvent('pxt.criticalerror', { error, info });
+        // Reload the page in 2 seconds
+        const lastCriticalError = pxt.storage.getLocal("lastcriticalerror") ?
+            Date.parse(pxt.storage.getLocal("lastcriticalerror")) : Date.now();
```
It doesn't but I can guard it anyway.

riknoll commented on webapp/src/app.tsx:
```diff
@@ -1760,6 +1782,15 @@ export class ProjectView
+            return <div id="root" className="ui middle aligned center aligned grid" style={{ height: '100%', alignItems: 'center' }}>
+                <div className="ui raised segment inverted purple">
+                    <h2>{lf("Oops")}</h2>
+                    {lf("We detected a problem and we will reload the editor in a few seconds..")}
```
Show a different message if it isn't going to reload? No idea what that message should say.

samelhusseini replied:
I think that’s over complicating the feature. Users really shouldn’t get into a reload loop, we just added a bit of logic to avoid that.

samelhusseini replied:
I think that’s over complicating the feature. Users really shouldn’t get into a reload loop, we just added a bit of logic to avoid the possibility
